### ISSUE TITLE
Please change default cursor placement after adding a tag or list

### ISSUE THREAD

**Issue by fmartelg** [mpcjanssen/simpletask-android#407]
Use case:

1. I have a list _+-waiting.for_ and a tag _@Fred_.
2. A typical entry might for me might read _+-waiting.for @Fred to send over report_.
3. Here is what happens when I try to type above entry:
    1. After inserting _+-waiting.for_ list cursor is placed at the beginning of the entry, so I have to shift it to end of line;
    2. After inserting _@Fred_ tag cursor is placed at the beginning of the entry, so I have to shift it to end of line;
    3. If I then insert a Priority _(A)_ cursor remains in place, which is what I would expect.

This workflow is very annoying.  I can see some users might want their lists and tags at the end of the entry, but IMHO that ought to be optional.

Thanks!

**Comment by smichel17** [comment]
From the use case you give, it seems like your problem isn't actually with adding lists and tags at the end of the list, but with where the cursor is placed after you do so.

If you add lists and tags at the cursor, instead of the end, someone can't edit a task, tap anywhere in the line, and then change tags/lists through the app's interface (because you could potentially insert tags in the middle of words, etc).

However, I don't think it breaks anything to place the cursor at the end of the line or leave it in place, when adding lists/tags.

**Comment by fmartelg** [comment]
Indeed, my problem is not adding tags or lists at the end of an entry.  However, this seems to be the default behaviour, and there is no way to change it.

> If you add lists and tags at the cursor, instead of the end, someone can't edit a task, tap anywhere in the line, and then change tags/lists through the app's interface (because you could potentially insert tags in the middle of words, etc).

Got it.

> However, I don't think it breaks anything to place the cursor at the end of the line or leave it in place, when adding lists/tags.

Correct, except it might conflict with the case where cursor is in middle of word and you add a tag.

What I want is for tag to be inserted where cursor is, and cursor left in position as if I had typed out the tag.  Almost like text autocompletion works. One might need to:

(a) leave it to users not to insert tags in middle of words;

(b) have an exception to handle such cases (e.g. tag or list placed at end, cursor left in place).

**Comment by smichel17** [comment]
>What I want is for tag or list to be inserted where cursor is, and cursor left in position as if I had typed out the tag or list.

I understand. I'm proposing a different solution that I think will solve your issue. I'm proposing the different solution because:

>(a) leave it to users not to insert tags in middle of words;

I would consider this a regression. When we can protect users from messing up, we should.

>(b) have an exception to handle such cases (e.g. tag or list placed at end, cursor left in place).

This is a good solution, but requires a lot more dev work.

---

Placing the cursor at line end after adding a list or tag should only require changing ~4 lines of code, and doesn't require any special cases / additional code complexity. It also means that the sequence you described at the top will work. If you can show me a different use case that isn't solved by placing the cursor at the end of the line, after the list/tag, I'll happily change my mind :)

**Comment by fmartelg** [comment]
I see. That works for me.

Thanks!

**Comment by smichel17** [comment]
(worth noting that while I've chimed in / done design work on issues here enough that I hope @mpcjanssen respects my opinion, it's still just my opinion and he might have a different one)

**Comment by mpcjanssen** [comment]
I will have to review the proposals in more detail, but a major limiting factor in all this is the absolute crap cursor handling Android provides. You can only retrieve and set the cursor position relative to the beginning of the text box, so for example line 3 column 5 is not supported.

Another difficulty is that it's easy to say:

> However, I don't think it breaks anything to place the cursor at the end of the line or leave it in place, when adding lists/tags.

But to calculate the actual cursor position is tricky because the contents have changed.

If you add to that the fact that you get a crash if you set an invalid position and you have a recipe for the current sub optimal scenario.

I personally always manage lists and tags from the main screen and not from the edit screen so you might consider that too.

But as said initially I will look into the different proposals.

**Comment by fmartelg** [comment]
Thanks Mark. BTW I remember we went through something similar when setting
a priority would place the cursor after the priority. The default now is to
leave it in place.

**Comment by mpcjanssen** [comment]
@fmartelg Priorities is a much easier problem because the cursor position is always -3, same or +3

**Comment by smichel17** [comment]
Oof. That is... interesting.

I looked quickly at AddTask.kt, and here's my current understanding of how this works; please correct me if I'm wrong.

>The text box is generated by concatenating all the tasks, with newline as a separator. Then when you add a tag (or list, but I'm tired of typing list/tag every time), the processing happens one tag at a time, at the task level, and then the text box for the gui is re-generated. I didn't see where the cursor position is set, but I'll just assume from your explanation that there's something like `set_cursor_position(int i)` and `int get_cursor_position()`.

After thinking through several implementations and realizing there's always annoying edge cases that don't work... I see where you're coming from, even leaving in place is hard. Well, doing it isn't hard. Finding an elegant solution that you don't want to exorcise from your code... that's hard.

Leaving in place is a slightly better option than moving to the beginning of the line, but it's a pain to implement and won't solve @fmartelg's issue anyway, since you'll still need to move the cursor to the end of the line after inserting lists/tags.

You have a way to put the cursor at the beginning of a given line. I presume there's a relatively painless way to figure out how many characters are in a line. It won't be 4 lines of code like I assumed, but it's totally consistent from both the developer and user perspectives and it'll solve the issue that spawned this ticket.

---

I'm currently thinking about a new interface (because I apparently really like giving @mpcjanssen work\* :stuck_out_tongue:) for entering lists, tags, and priorities -- one that resembles autocompletion -- but who knows how long I will be thinking about that for, so don't hold off on this change on my account.

_\* Actually my hope is that I can take design work off your hands, so you can focus on implementation. Let me know if there's any areas you're particularly thinking about revamped designs for._

**Comment by smichel17** [comment]
@fmartelg would implementing my suggestion on #182 (auto-completion while typing) be sufficient to close this ticket?

**Comment by mpcjanssen** [comment]
It would be, but see my remark in #182
